### PR TITLE
Remove hand rolled hashing

### PR DESCRIPTION
**Pull request by systay**

### CODE REVIEW

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala [hunk 1]
```diff
@@ -90,3 +93,29 @@ case class DistinctSlottedPrimitivePipe(source: Pipe,
...
   }
 }
+
+/**
+  * This little class is here to make sure we have the expected behaviour of our primitive arrays.
+  * In the JVM, long[] are do not have reasonable hashcode or equal()
+  */
+class Key(val inner: Array[Long]) {
+
+  private val _hashCode = util.Arrays.hashCode(inner)
```
Comment (pontusmelke) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala]: Couldn't you just do `override val hashCode = util.Arrays.hashCode(inner)`?

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala [hunk 2]
```diff
@@ -90,3 +93,29 @@ case class DistinctSlottedPrimitivePipe(source: Pipe,
...
+
+  override def equals(other: Any): Boolean = {
+    if ( other == null || getClass() != other.getClass() )
```
Comment (pontusmelke) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala]: I think this is equivalent with
```
other match {
   case k: Key => _hashCode == otherKey._hashCode && util.Arrays.equals(inner, otherKey.inner)
   case _ => false
}
```

I don't think that will create any lambdas just do an instance of check, I'm always scared about using return.

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/EagerAggregationSlottedPrimitivePipe.scala [hunk 3]
```diff
@@ -79,20 +78,20 @@ case class EagerAggregationSlottedPrimitivePipe(source: Pipe,
       }
     }
 
-    val supplier: Supplier[Seq[AggregationFunction]] = new Supplier[Seq[AggregationFunction]] {
-      override def get(): Seq[AggregationFunction] = aggregationFunctions.map(_.createAggregationFunction)
+    val supplier = new java.util.function.Function[Key, Seq[AggregationFunction]] {
```
Comment (pontusmelke) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/EagerAggregationSlottedPrimitivePipe.scala]: maybe rename now that it is no longer a `Supplier`?

**Review by pontusmelke — APPROVED** [review]
Slightly sad to see the `LongArrayX` go but I guess -1000 lines of code is nice. I just had some minor stuff - none of them is very important.

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala (hunk 1 above)
Comment (systay) [reply enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala]: You are probably right.

**Pull request merged**